[feature] Enhances Garrison with new system features

loxleygames opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?
These system enhancements are cherry-picked from [AirSkyBoat](https://github.com/AirSkyBoat/AirSkyBoat/pulls?q=is%3Apr+garrison+is%3Aclosed). The original work was done by @xkoredev, who authored the commits. I've tested the features in game and verified the system is working correctly.

I'm very keen on working towards completing this content on LandSandBoat and willing to follow through with tidying or polish to get these features up to the project's satisfaction.

### Overview of the features implemented

* Adds debugging for Garrison events (@xkoredev)
* Garrison data moved to new global `globals/garrison_data.lua` (@xkoredev)
* Merged existing capture data and collected some new data (@LoxleyXI)
* Refactoring and tidy up work to meet the style standards (@LoxleyXI )
* Disable Garrison in areas where the capture is incomplete (@LoxleyXI)
* Adds optional module to provide placeholder data, and NPC combatants with a randomised weapon if  missing (Modularised by @LoxleyXI, original work and placeholders by @xkoredev and contributors from [AirSkyBoat](https://github.com/AirSkyBoat/AirSkyBoat))

#### Commands
Adds commands for development and testing (@xkoredev):
* * !garrison start (player) starts the garrison for the given player (or targetted one). This bypasses requirements like lockout
* * !garrison stop  (player) stops the garrison (if any) currently running in the player's zone
* * !garrison win (player) win the garrison (if any) currently running in the player's zone

#### Battle States
Adds the following new states (@xkoredev):
* * `BATTLE` - Manages the spawning of waves or finishing the event (ie. all NPCs dead)
* * `SPAWN_BOSS` - Manages spawning the boss and gracefully handling errors (eg. missing mob)
* * `ADVANCE_WAVE` - Sets the delay between waves
* * `SPAWN_MOBS` - Spawn the correct number of mobs and make them aggro NPCs
* * `GRANT_LOOT` - Handle success message, rolling out loot and distributing gil

#### Functions
Adds the following new functions (@xkoredev):
* * `xi.garrison.addLevelCap` - Replaces local `addLevelCap`, making it accessible to commands, and adds flag for `DEATH`
* * `xi.garrison.aggroGroups` - Replaces local `aggroGroups` and adds logging
* * `xi.garrison.spawnNPC` - Replaces local `spawnNPC`, adds `GARRISON_NPC_DEATH` listener to track lose conditions, and uses `GM_HOME` to store the groupId data for the dynamic entities
* * `xi.garrison.spawnNPCs` - Spawn 1 NPC per player, setting the appropriate name, look and group ID where captured
* * `xi.garrison.spawnMob` - Spawn mobs under level cap and add listeners to handle the win/wave conditions
* * `xi.garrison.pickMobsFromPool` - Randomly select mobs from the pool for the current wave
* * `xi.garrison.watchdog` - Ticks at an interval to ensure the current Garrison state is valid, and cleans up any lingering Garrison events
* * `xi.garrison.stop`- Clean up current battle, removing level restrictions, despawning mobs and NPCs
* * `xi.garrison.handleLootRolls` - Roll random loot from the loot table among all players
* * `xi.garrison.handleGilPayout` - Splits a total gil payout between the players present
* * `xi.garrison.getSpawnSchedule` - Gets spawn schedule for the current zone
* * `xi.garrison.validateEntry` - Check lockout for the current player based on the Garrison settings. For example, the nation must control the outpost, the player must be a minimum rank, and the player can only complete one Garrison per tally. All of these can be modified in the settings

#### Lockouts
Adds the following lockout handling functions (@xkoredev):
* * `xi.garrison.saveTallyLockout` - Updates player's `[Garrison]NextEntryTime` var unless `GARRISON_ONCE_PER_WEEK` is turned off in settings
* * `xi.garrison.isPlayerOnTallyLockout` - Return true if player has remaining duration on `[Garrison]NextEntryTime` unless `GARRISON_ONCE_PER_WEEK` is turned off in settings. Prints remaining time if debugging is turned on in settings
* * `xi.garrison.saveZoneLockout` - Update's players lockout time using the current lockout settings
* * `xi.garrison.isZoneOnLockout` - Return if player has remaining duration on their lockout for the current zone (`[Garrison]NextEntryTime_ZONEID`)
* * `xi.garrison.win` - Used to trigger the win condition from commands

### Regarding the state of capture data
This PR includes some previously missing capture data. The whole set of data has not been provided due to the difficulty of obtaining it. I strongly encourage future developers to continue this work, so that the additional areas can be unlocked in base.

Some of the challenges of obtaining Garrison data are as follows:

* Players can only participate in Garrison once per week
* The player's nation must control the current region. This may be practically impossible for some region/nation combos depending on the server.
* The number of NPCs spawned is based on the number of real players (Guaranteeing a complete capture would mean getting 18 players to do this obsolete content, good luck with that)
* There are 15 different nation/level/region combinations that each need to meet these very tough requirements, that can only be captured one week at a time.

## Steps to test these changes
Ronfaure example:
* `!setplayernation (name) 0` (San d'Oria)
* `!setrank (name) 2`
* `!additem red_cryptex`
* Trade to guard (eg. Doladepaiton, R.K.)

xkoredev committed:
Implement Garrison System

- Implements wave system that spawns the right amount of mobs at the right time for each wave
- Fixes the number of allies spawned (1 per alliance member based on captures)
- Grants correct loot on victory
- Spawns the right bosses / npcs on each zone
- Save lockout information for player
- Add timer for garrison timeout
- Gates entry based on rank and other requirements

xkoredev committed:
Garrison: Add watchdog and gm command polish

- Adds a timer to poll for inconsistent garrison state and clean up
- Change stopgarrison to !garrison stop and !garrison start commands

xkoredev committed:
Polish Garrison npcs

- Fixes ally npc looks
- Assigns proper group id

xkoredev committed:
Fix garrison timer lockouts according to spec (#2110)

panicstevenson committed:
Garrison Drops Into Group

panicstevenson committed:
Garrison Drops Into Group

xkoredev committed:
Fix garrison waves (#2203)

- Fixes wave spawn schedules to match those from captures

xkoredev committed:
Fix garrison npc levels and loot tables

- Fixes garrison npc levels so they can be adjusted with group ids
- Reduces mannequin drop rates to 5%
- Adds logging when dynamic entity cannot be found

xkoredev committed:
Add garrison gil payout

xkoredev committed:
Add random weapon models for garrison allies

loxleygames committed:
Removes enum and print

zach2good commented on src/map/utils/mobutils.cpp:
```diff
@@ -1527,6 +1527,11 @@ a mobs regen rate, magic defense, triple attack rate from a table instead of har
         }
+        else
+        {
+            ShowError("Unable to find entity with groupId: %d, zoneId: %d. Check that mob_pools.ele_eva_id, group and zoneid match.", groupid, groupZoneId);
```
LSB doesn't have `mob_pools.ele_eva_id`

zach2good commented on scripts/globals/garrison_data.lua:
```diff
@@ -0,0 +1,673 @@
+{
+    [20] = { "Patrician", "Recruit", "Candidate" },
+    [30] = { "Trader", "Mariner", "Scholar" },
+    -- TODO: Capture Windy npc. Shouldn't be TempleKnight.
```
👀

zach2good commented on scripts/globals/garrison_data.lua:
```diff
@@ -0,0 +1,673 @@
+xi.garrison.allyLooks =
+{
+    -- Level 20 garrison looks
+    -- The looks that have not been obtained from captures are commented
```
👀

loxleygames replied:
I will do what I can to obtain the missing data, but there is a lot. Is there any room for  compromise, such as blocking the areas that are missing captures, so at least some of the areas are usable? If not, I think requiring this to be complete before accepting any enhancements is an impossibly high bar and it will probably never be done.

This is one of the earliest pieces of content that still isn't properly supported. I think finishing Garrison would go a long way to completing emulation for the base game. I'm also willing to refactor what's here into a form that you prefer.

zach2good replied:
> as blocking the areas that are missing captures

Yeah, this will be fine.

> An impossibly high bar

Can I remind you that you've opened a PR called `Implements Garrison`, using 10 commits of previous contributors' work? If you don't want to scope the whole feature, change the title and just submit a single confirmed-working Garrison at a time - alongside any system and core changes needed as you go.

The reason I'm so aggressive about this is incorrect work gets fixed/finished _even less_ than partially completed work. If it's a feature that not many people care about and it's a vague approximation of what it's supposed to be - then it'll be left in that state forever. While you're here with the code fresh in your mind and an understanding of the feature - that's the perfect time to actually implement it properly.

Also, upstream holds itself to higher quality and correctness standards than downstream - because everyone builds off of us. More servers and groups than just ASB and Horizon use it, and it's unfair to them to just yeet half-finished work into the repo.

loxleygames replied:
Thanks for your feedback, which I genuinely appreciate. I was not expecting to push things in the original state, and your comments allowed me to more easily identify the areas for improvement.

I hope I've alleviated some concerns by properly documenting the features and moving custom behavior to a module. Regarding code, I agree on high standards of correctness and quality without hesitation, the bar referred only to the completeness of captures, which are challenging to obtain for this content (One of the reasons, it's still so bare).

With this enhancements to this system, I'm optimistic the missing capture data will begin to be submitted, as it will unlock those areas for immediate use on live servers.

If there are any further concerns about the implementation, I will gladly look into resolving them.

TeoTwawki replied:
I'm optimistic that models aren't difficult to acquire from people with retail access before hand.

I wish I had a cap suite that worked on ashita instead of widower, I'd go suicide on garrison for you all week long For now I've made a request and on [captures discord](https://discord.com/channels/443544205206355968/443893540922064896/1133447193441865778), where I see you are already a member.

Also this is a thing, but almost nothing is properly labelled: https://github.com/LandSandBoat/lsb-wiki/blob/main/Model-ID-Reference.md

I've made [an appeal](https://discord.com/channels/392903136336936960/883227930828886060/1133184440688984085) to the wider community hoping to find people interested in fixing that up, but I am less optimistic about that than I am captures.

> he bar referred only to the completeness of captures, which are challenging to obtain for this content (One of the reasons, it's still so bare).

You kinda set that bar yourself at 1st (per Zach's reply, doesn't all have to happen at once), though I do not believe that was "impossibly" high. If you want to do parts label your pr and commits for your parts if it doesn't "work" without missing parts that's ok.\* If need be messages can be edited while still retaining author info.

<sub><sup>\*</sup> if something "sorta works" even when its wrong ppl are less encouraged to fill it in correctly _and_ the proj gets blamed for it being wrong by the ppl that actually pay attention to that. I could write and entire angry editorial about how many times I've seen the teams (plural, multiple iterations since dsp but to some people we're "all the same") get disparaging comments for it which resulted in the current crafting table and its missing crafts after a previous project lead had enough of that. </sub>

LoxleyXI replied:
> models aren't difficult to acquire from people with retail access before hand

While I respect the sentiment, here are the reasons why Garrison captures are especially challenging:
* Players can only participate in Garrison once per week
* The player's nation *must* control the current region. This may be practically impossible for some region/nation combos depending on the server.
* The number of NPCs spawned is based on the number of *real* players (Guaranteeing a complete capture would mean getting 18 players to do this obsolete content, good luck with that)
* There are 15 different nation/level/region combinations that each need to meet these very tough requirements, that can only be captured one week at a time.

Outside of Chocobo raising, this is probably one of the worst pieces of content capture wise. And at least you can solo Chocobo raising (Though appreciate there's more content *to* capture with raising). To get the most productive captures, you can't solo Garrison.

TeoTwawki replied:
gosh I has no idea! I am SO shocked!

-_-

Do you really want me to explain to you why retorting like this is a bad idea?

zach2good replied:
> The number of NPCs spawned is based on the number of real players (Guaranteeing a complete capture would mean getting 18 players to do this obsolete content, good luck with that)

This one we haven't proven yet (#shitwikisays, etc.). We think it's a function of both total players (Trusts can't be called, unsure about Adventuring fellows) and the number of parties involved. Is a 6-player party going to spawn as many waves of certain sizes as 3x parties of 2? etc.

loxleygames replied:
That's a good point. If I can get a few people to help test as an alliance (1 per party), I will make the adjustments to that function if required.

TeoTwawki replied:
> > The number of NPCs spawned is based on the number of real players (Guaranteeing a complete capture would mean getting 18 players to do this obsolete content, good luck with that)
> 
> This one we haven't proven yet (#shitwikisays, etc.). We think it's a function of both total players (Trusts can't be called, unsure about Adventuring fellows) and the number of parties involved. Is a 6-player party going to spawn as many waves of certain sizes as 3x parties of 2? etc.

I know it _used to_ just count parties (I had tested it about 2011 or so) and also SE rarely updates older content without some announcement of it (like stripping level caps) so its a good possibility its  still working that way. 

Alas, if my system had good capture tools (ashita..) and I had enough days off it I could also nation change and cap regions one per week as needed (I can drum up 17 warm bodies easy enough) as long as the win event data isn't needed on a max size alliance (17 winners I can _not_ drum up lol)

TeoTwawki replied:
worth noting more modern advanced jobs (blu, pup, run) likely go a long way on these as the content was designed before they existed and wouldn't have been balanced with them in mind.

zach2good reviewed (commented):
It's been a long time since I've looked at this code, but I thought the current state of this was already in the codebase 🤷 

I remember blocking this because we still need a bunch of caps - it's functionally incomplete. How often do people run this code on ASB and Horizon? It wasn't correct or stable last time I looked if memory serves

loxleygames committed:
Fixes formatting and error message

loxleygames committed:
Improves table alignment, adds better spacing

loxleygames committed:
Prevents starting Garrison with missing capture

Provides message to players
Handles !garrison commands gracefully
Placeholder NPCs moved to custom module
addWeaponIfNecessary moved to custom module

loxleygames committed:
Improves debug message, move tables to ID

loxleygames committed:
Adds npc name for Windurst 40

loxleygames committed:
Adds NPC looks for 20 and 30

loxleygames committed:
Merge branch 'base' into garrison-asb-cherrry-pick

LoxleyXI commented:
Assuming you are not expecting complete capture data, is there anything else you guys are expecting from this PR?

zach2good commented on scripts/globals/garrison_data.lua:
```diff
@@ -0,0 +1,622 @@
+        textRegion = 1,
+        levelCap = 20,
+        mobBoss = "Lead_Quadav",
+        xPos = -580, --needs adjusting
```
Does this still need adjusting?

zach2good commented on scripts/globals/garrison_data.lua:
```diff
@@ -0,0 +1,622 @@
+        textRegion = 2,
+        levelCap = 20,
+        mobBoss = "Yagudo_Condottiere",
+        xPos = -20, -- needs adjusting
```
Does this still need adjusting?

zach2good commented on scripts/globals/garrison_data.lua:
```diff
@@ -0,0 +1,622 @@
+        textRegion = 3,
+        levelCap = 30,
+        mobBoss = "Goblin_Swindler",
+        xPos = 141, -- needs adjusting
```
Does this still need adjusting?

zach2good commented on scripts/globals/garrison_data.lua:
```diff
@@ -0,0 +1,622 @@
+        textRegion = 4,
+        levelCap = 30,
+        mobBoss = "Orcish_Colonel",
+        xPos = 54, -- needs adjusting
```
Does this still need adjusting?

zach2good commented on scripts/globals/garrison_data.lua:
```diff
@@ -0,0 +1,622 @@
+        textRegion = 5,
+        levelCap = 30,
+        mobBoss = "Cobalt_Quadav",
+        xPos = 458, -- needs adjusting
```
Does this still need adjusting?

loxleygames replied:
I went through all the available captures and reviewed each of these.
Have moved unknown / missing positions to the placeholder module.
Also better aligned these tables to fit LSB standards.

zach2good commented on scripts/globals/garrison_data.lua:
```diff
@@ -0,0 +1,622 @@
+        textRegion = 6,
+        levelCap = 30,
+        mobBoss = "Goblin_Guide",
+        xPos = -485, -- needs adjusting
```
Does this still need adjusting?

zach2good commented on scripts/globals/garrison_data.lua:
```diff
@@ -0,0 +1,622 @@
+        textRegion = 7,
+        levelCap = 30,
+        mobBoss = "Yagudo_Missionary",
+        xPos = -299, -- needs adjusting
```
Does this still need adjusting?

zach2good commented on scripts/globals/garrison_data.lua:
```diff
@@ -0,0 +1,622 @@
+        textRegion = 10,
+        levelCap = 30,
+        mobBoss = "Hunting_Chief",
+        xPos = -247, -- needs adjusting
```
Does this still need adjusting?

zach2good commented on scripts/globals/garrison_data.lua:
```diff
@@ -0,0 +1,622 @@
+        textRegion = 8,
+        levelCap = 40,
+        mobBoss = "Gigas_Overseer",
+        xPos = -25, -- needs adjusting
```
Does this still need adjusting?

zach2good commented on scripts/globals/garrison_data.lua:
```diff
@@ -0,0 +1,622 @@
+        textRegion = 9,
+        levelCap = 50,
+        mobBoss = "Demon_Aristocrat",
+        xPos = 216, -- needs adjusting
```
Does this still need adjusting?

zach2good commented on scripts/globals/garrison_data.lua:
```diff
@@ -0,0 +1,622 @@
+    {
+        itemReq = xi.items.BUNNY_FANG_SACK,
+        textRegion = 13,
+        levelCap = 75, -- 75 Mobs/NPCs (Uncapped)
```
Does this comment need to be here?

loxleygames replied:
Changed level cap to 99 and pull from `xi.settings.main.MAX_LEVEL` in this case

zach2good commented on settings/default/main.lua:
```diff
@@ -167,6 +167,16 @@ xi.settings.main =
     -- Multiplier to NM lottery cooldown time (Default 1.0) eg. 2.0 = twice as long. 0 = no cooldowns.
     NM_LOTTERY_COOLDOWN = 1.0,
 
+	-- GARRISON SETTINGS
```
This is indented with a tabm should be 4x spaces

zach2good commented on settings/default/main.lua:
```diff
@@ -167,6 +167,16 @@ xi.settings.main =
+    GARRISON_TIME_LIMIT          = 1800,   -- Time in seconds before lose ongoing garrison (default: 1800)
+    GARRISON_ONCE_PER_WEEK       = false,  -- Set to false to bypass the limit of one garrison per Conquest Tally Week.
+    GARRISON_PARTY_LIMIT         = 18,     -- Set to max party members you want to do garrison (default: 18).
+    GARRISON_NATION_BYPASS       = false,  -- Set to true to bypass the nation requirement.
```
These seem arbitrarily aligned out? Just one space past the longest key for everything in this section

TeoTwawki replied:
additionally settings should _**default**_ to whatever modern retail does, even if we know pretty much everyone will change them.

loxleygames replied:
Fixed and corrected defaults

zach2good commented on src/map/utils/mobutils.cpp:
```diff
@@ -1527,6 +1527,11 @@ a mobs regen rate, magic defense, triple attack rate from a table instead of har
         }
+        else
+        {
+            ShowError("Unable to find entity with groupId: %d, zoneId: %d.", groupid, groupZoneId);
```
This has been deployed on Horizon for a while right? This doesn't trigger noisily from other or custom content?

loxleygames replied:
Happy to remove this if not wanted

zach2good commented on scripts/globals/garrison_data.lua:
```diff
@@ -0,0 +1,622 @@
+            "0x0100020260102420603060406050B56000700000",
+            "0x010008083D1024203D3010401050756075700000",
+            "0x01000008371024203730374037506F6018700000",
+            "0x0100040105102420053005400550BB6000700000"
```
Final entries should have a trailing comma too

zach2good commented on scripts/globals/garrison_data.lua:
```diff
@@ -0,0 +1,622 @@
+        [xi.nation.SANDORIA] =
+        {
+            "0x01000E04191019201930194019506B601C700000",
+            "0x01000903191019201930194019506B601C700000"
```
Final entries should have a trailing comma too

loxleygames replied:
Fixed all of these

zach2good commented on scripts/globals/garrison_data.lua:
```diff
@@ -0,0 +1,622 @@
+            "0x01000B051C1073201430144014506C6000700000",
+            "0x0100010777106920693066406950B46000700000",
+            "0x010000067D107820033066406850E96000700000",
+            "0x0100020700100220023002400250006000700000", -- Naih_Arihmepp
```
Is this correct? I thought you stripped out guessed looks from NPCs?

loxleygames replied:
Missed this one in the move. Have removed and re-checked all the NPCs.

zach2good commented on scripts/globals/garrison_data.lua:
```diff
@@ -0,0 +1,622 @@
+            "0x010006030010762076303A400650736000700000",
+            "0x01000F0300101520153015401550006000700000",
+            "0x010009040010762076303A400650736000700000",
+            "0x01000E0400101520003015401550006000700000"
```
Naih_Arihmepp

zach2good commented on scripts/globals/garrison_data.lua:
```diff
@@ -0,0 +1,622 @@
+        { itemid = xi.items.ACCURATE_EARRING,  droprate =  350 },
+    },
+    --level 75 garrison loot
+    --TODO: 75 era module
```
Whats left TODO here?

loxleygames replied:
Multiple sources confirmed Cape Terrigan / Vollbow should be "uncapped" (99).
Changed all `75` references to `99`, and now defer to `main.MAX_LEVEL` in server settings.

TeoTwawki replied:
uncapped and "capped at max level" are 2 diff things, be sure use corrected terminology to avoid confusion. Capped at max level will strip buffs, uncapped won't. (I expect you have the right thing and just had the wrong wording)

zach2good reviewed (commented):
Whats the balance like on this? Is it comparable as a single person to retail? Mainly aiming for this not to be loot pinata

loxleygames committed:
Merge remote-tracking branch 'upstream/base' into garrison-asb-cherrry-pick

loxleygames committed:
Moves unknown positions to placeholder module

loxleygames committed:
Uncapped Garrison should be 99 by default

loxleygames committed:
Fixed settings alignment and defaults

LoxleyXI commented:
> Whats the balance like on this? Is it comparable as a single person to retail? Mainly aiming for this not to be loot pinata

Test example: As a level 20 Monk with average gear, I was able to kill Yagudo Votary in 90 seconds. The damage I took was not sustainable for a solo player (Around 10% of HP per hit), which is to be expected from group content. This seems to be an accurate level of difficulty.

zach2good commented:
And that's what retail is like?

loxleygames committed:
Moves position to allyInfo, clean up spawnNPCs

LoxleyXI commented:
Reviewed a few videos of the same Garrison and Yagudo Votary were taking 1.5-2mins to kill each and hitting for similar or less damage. Would say these are about right.

loxleygames committed:
Merge branch 'base' into garrison-asb-cherrry-pick

zach2good reviewed (approved):
^ Once this passes CI, that is.

Reviewers note: For once, these commits are all mostly sane. So as long as the next commit isn't "Remove enums to appease CI", then it should be fine to go un-squashed. Especially since 3x people have touched it.

Worst case scenario we can merge into a feature branch and do the squashing ourselves.

zach2good commented:
> Reviewed a few videos of the same Garrison and Yagudo Votary were taking 1.5-2mins to kill each and hitting for similar or less damage. Would say these are about right.

Are those the videos from the caps discord, or random garbage on YT? Is that also just testing a single Garrison? It's possible to go in and solo any/all of them, get your ass kicked, and still get some numbers/times you can test against.

LoxleyXI commented:
> Are those the videos from the caps discord, or random garbage on YT?

I've looked at everything, to get as much info as possible. For that case specifically (Balance / damage numbers) it was videos from the caps discord.

loxleygames committed:
Removed deprecated requires
